visualization/panda/panda3d_utils.py

```diff
@@ -0,0 +1,239 @@
+    # Resize and pad image while meeting stride-multiple constraints
+    shape = img.shape[:2]  # current shape [height, width]
+    if isinstance(new_shape, int):
+        new_shape = (new_shape, new_shape)
+
+    # Scale ratio (new / old)
+    r = min(new_shape[0] / shape[0], new_shape[1] / shape[1])
+    if not scaleup:  # only scale down, do not scale up (for better val mAP)
+        r = min(r, 1.0)
+
+    # Compute padding
+    ratio = r, r  # width, height ratios
+    new_unpad = int(round(shape[1] * r)), int(round(shape[0] * r))
+    dw, dh = new_shape[1] - new_unpad[0], new_shape[0] - new_unpad[1]  # wh padding
+    if auto:  # minimum rectangle
+        dw, dh = np.mod(dw, stride), np.mod(dh, stride)  # wh padding
+    elif scaleFill:  # stretch
+        dw, dh = 0.0, 0.0
+        new_unpad = (new_shape[1], new_shape[0])
+        ratio = new_shape[1] / shape[1], new_shape[0] / shape[0]  # width, height ratios
+
+    dw /= 2  # divide padding into 2 sides
+    dh /= 2
+
+    if shape[::-1] != new_unpad:  # resize
+        img = cv2.resize(img, new_unpad, interpolation=cv2.INTER_LINEAR)
+    top, bottom = int(round(dh - 0.1)), int(round(dh + 0.1))
+    left, right = int(round(dw - 0.1)), int(round(dw + 0.1))
+    img = cv2.copyMakeBorder(img, top, bottom, left, right, cv2.BORDER_CONSTANT, value=color)  # add border
+    return img, ratio, (dw, dh)
+
+
+class ImgOnscreen():
+    """
+    Add a on screen image in the render 2d scene of Showbase
+    """
+
+    def __init__(self, size, parent_np=None):
+        """
+        :param size: (width, height)
+        :param parent_np: Should be ShowBase or ExtraWindow
+        """
+        self._size = size
+        self.tx = Texture("video")
+        self.tx.setup2dTexture(size[0], size[1], Texture.TUnsignedByte, Texture.FRgb8)
+        # this makes some important setup call
+        # self.tx.load(PNMImage(card_size[0], card_size[1]))
+        self.onscreen_image = OnscreenImage(self.tx,
+                                            pos=(0, 0, 0),
+                                            parent=parent_np.render2d)
+
+    def update_img(self, img: np.ndarray):
+        """
+        Update the onscreen image
+        :param img:
+        :return:
+        """
+        if img.shape[2] == 1:
+            img = img_to_n_channel(img)
+        resized_img = letterbox(img, new_shape=[self._size[1], self._size[0]], auto=False)[0]
+        self.tx.setRamImage(resized_img.tostring())
+
+    def remove(self):
+        """
+        Release the memory
+        :return:
+        """
+        if self.onscreen_image is not None:
+            self.onscreen_image.destroy()
+
+    def __del__(self):
+        self.remove()
+
+
+class ExtraWindow():
+    """
+    Create a extra window on the scene
+    TODO: small bug to fix: win.requestProperties does not change the properties of window immediately
```

Comment:
Is this bug fixed?

Reply:
No, I am not sure that this is a bug or feature. The `window properties` changes only after base.run(). Before base.run, the window properties remain unchanged.